Implement SmallArray

harrishancock opened the pull request:
These utilities originally landed in https://github.com/capnproto/capnproto/pull/1762. Unfortunately, they landed at the same time as a [subtle bug in our fiber implementation](https://github.com/capnproto/capnproto/pull/1842) began causing production crashes in the Workers runtime, and we reverted this `SmallArray` feature out of an abundance of caution.

I've been meaning to re-submit `SmallArray`. It is a more ergonomic interface than KJ_STACK_ARRAY, and the primary compiler we target nowadays (clang) does not support VLAs, which is KJ_STACK_ARRAY's competitive advantage.

In the intervening years, some HeapArrayDisposer was refactored to use `if constexpr` instead of a dispatch struct template in: https://github.com/capnproto/capnproto/pull/2288/files#diff-a50b82e04212952a2c02eee72504e7cca446a9b6bc0c89348bdf3f34fa5b4fa0. This meant that my old PR did not cleanly apply, but ultimately it made the diff here easier to follow, obviating a preliminary refactoring commit.

In my original PR where this work landed, I replaced KJ_STACK_ARRAY wholesale in a synchronized multiple-repo. That was, in retrospect, dumb and unnecessary. I would like to go through and replace instances of KJ_STACK_ARRAY with `SmallArray` and `SmallArrayBuilder`, but that can come at our leisure.

harrishancock commented on c++/src/kj/array.h:
```diff
@@ -694,10 +807,28 @@ void ArrayDisposer::dispose(T* firstElement, size_t elementCount, size_t capacit
 
 template <typename T>
-struct HeapArrayDisposer::Allocate_ {
+struct Allocate_ {
```
I had to un-nest this helper.

harrishancock commented on c++/src/kj/array.h:
```diff
@@ -722,6 +853,24 @@ T* HeapArrayDisposer::allocateUninitialized(size_t count) {
 }
 
+template <typename T>
+T* SmallArrayDisposer::allocate(void* firstElement, size_t count) {
```
These two functions are now largely just copies of their HeapArrayDisposer counterparts. In my earlier PR, I had them all share code -- but the code they shared was the template metaprogramming version of `if constexpr`, and it's much clearer to just duplicate that, now.

harrishancock commented:
Interesting, I stand corrected, then. I think I must have assumed clang didn't support them because our KJ_STACK_ARRAY implementation only used them on GCC, but perhaps that's just because GCC was the primary compiler in capnp's early years.

kentonv commented on c++/src/kj/array.h:
```diff
@@ -144,7 +144,7 @@ class Array {
 
   KJ_DISALLOW_COPY(Array);
-  inline ~Array() noexcept { dispose(); }
+  inline ~Array() noexcept(false) { dispose(); }
```
Yeah this seems like a (very longstanding) bug.

kentonv commented on c++/src/kj/array.h:
```diff
@@ -316,9 +316,26 @@ class HeapArrayDisposer final: public ArrayDisposer {
+
+private:
+  static void* allocateImpl(void* firstElement, size_t elementSize, size_t elementCount,
+                     size_t capacity, void (*constructElement)(void*),
```
Hmm, I wonder if having to virtualize the element constructor/destructor even in the small-buffer case will prove to be a significant performance hit vs. KJ_STACK_ARRAY.

With KJ_STACK_ARRAY we are assuming the type always has a default constructor and that it's OK for us to construct more elements than we actually use. Construction is probably pretty cheap in most if not all cases (probably just zeroing), though I suppose destruction could be doing a lot of wasteful branches checking if things are still null...

kentonv replied:
I'm not sure we actually need the full logic of `HeapArrayDisposer`. If we simply require that the type has a default constructor, then we can allocate the small array as a static-sized array with all elements constructed upfront (which is what KJ_STACK_ARRAY does). In this case construction and destruction are just handled by the compiler and we don't need a disposer at all.

harrishancock commented on c++/src/kj/array.c++:
```diff
@@ -105,5 +113,24 @@ void HeapArrayDisposer::disposeImpl(
 
 const HeapArrayDisposer HeapArrayDisposer::instance = HeapArrayDisposer();
 
+void* SmallArrayDisposer::allocateImpl(
```
Any specific ideas? I'm not sure I follow.

harrishancock commented:
Updated with a simpler implementation which always default-constructs the inline array, like KJ_STACK_ARRAY does. I dropped SmallArrayBuilder, meaning HeapArrayDisposer is now undisturbed. Additionally, I added `first()` and `as()` support to SmallArray, which were added since my previous PR. Lastly, I fleshed out the commit message a bit more.
